File: ios/ParkingKok/Detection/Trace/TraceLabelPromptNotifier.swift
import Foundation
import UserNotifications

/// What `TraceRecorder` is allowed to know about prompting for a label.
///
/// One method, no return value, never throws: a closed session is already on disk when
/// this is called, and a notification that cannot be posted must not be able to break the
/// recording path (docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §9 best-effort).
protocol TraceLabelPrompting: Sendable {
    /// Fire-and-forget. Whether anything is actually shown is the implementation's
    /// business, including deciding that permission is missing.
    func requestPrompt(_ prompt: TraceLabelPrompt)
}

/// The notification system, reduced to the two things prompting needs.
///
/// Exists so the decision in `TraceLabelPrompter` — post, or count a suppression — is
/// testable. `UNUserNotificationCenter.current()` cannot be reached from a unit test
/// without the real notification service behind it.
protocol LabelPromptDelivering: Sendable {
    /// Whether the app may show an alert at all. `false` covers both "never asked" and
    /// "user said no"; neither is an error and neither is retried.
    func isAuthorized() async -> Bool
    /// Registers the actions and posts. Swallows its own failures — see the protocol above.
    func deliver(_ prompt: TraceLabelPrompt) async
}

/// The names the notification and the tap that comes back have to agree on.
///
/// Its own category identifier, never `parking_detection` or the M3 candidate category:
/// this asks about a trip that already ended, and a diagnostics prompt sharing a channel
/// with a product notification would make "the user muted the wrong one" unanswerable.
enum TraceLabelPromptAction {
    static let categoryIdentifier = "pk.diagnostics.trace_label_prompt"
    static let sessionIdKey = "pk.diagnostics.trace_session_id"
    private static let actionPrefix = "pk.diagnostics.trace_label."

    static func identifier(for mode: TraceMode) -> String {
        actionPrefix + mode.rawValue
    }

    /// `nil` for anything that is not one of our label actions — the body tap
    /// (`UNNotificationDefaultActionIdentifier`), a dismissal, or an identifier left over
    /// from a build that offered a different set.
    static func mode(forActionIdentifier identifier: String) -> TraceMode? {
        guard identifier.hasPrefix(actionPrefix) else { return nil }
        return TraceMode(rawValue: String(identifier.dropFirst(actionPrefix.count)))
    }
}

/// Decides whether a closed session is worth a notification, and counts it when it is not.
///
/// The decision is the whole of this type, which is why it is separate from the delivery it
/// drives: a missing permission has to be *visible* rather than silent, because a field
/// weekend that collected nothing would otherwise look exactly like a weekend nobody
/// travelled (§9 "조용히 버리지 마라").
struct TraceLabelPrompter: TraceLabelPrompting {
    private let delivery: any LabelPromptDelivering
    /// Bumps the persisted counter the diagnostics report reads. A closure rather than the
    /// store itself: this needs one number incremented, not a trace store.
    private let onSuppressed: @Sendable () -> Void

    init(delivery: any LabelPromptDelivering, onSuppressed: @escaping @Sendable () -> Void) {
        self.delivery = delivery
        self.onSuppressed = onSuppressed
    }

    /// Hands the work to an unstructured `Task`, because the recorder's callers are
    /// synchronous and a session must be considered closed the instant its file is.
    ///
    /// Unstructured on purpose: the enclosing task is a detection callback that may finish
    /// immediately, and a child task would be cancelled with it. The cost is that a process
    /// suspended in the next instant loses the prompt — acceptable for instrumentation, and
    /// the alternative would be making the recording path wait on the notification service.
    func requestPrompt(_ prompt: TraceLabelPrompt) {
        Task { await resolvePrompt(prompt) }
    }

    /// The same decision, awaited. `requestPrompt` is exactly this plus a `Task`, so a test
    /// asserts on the outcome instead of racing an unstructured task.
    func resolvePrompt(_ prompt: TraceLabelPrompt) async {
        guard await delivery.isAuthorized() else {
            onSuppressed()
            return
        }
        await delivery.deliver(prompt)
    }
}

/// Posts through `UNUserNotificationCenter`.
///
/// Four actions, which is what iOS shows on an expanded notification; the ordering in
/// `TraceLabelPrompt.offeredModes` is what decides who gets cut if that ever shrinks.
/// `.foreground` is deliberately *not* set on any of them — the point of this whole change
/// is that labelling costs one tap and never an app launch.
struct UserNotificationLabelPromptDelivery: LabelPromptDelivering {
    /// iOS shows at most four actions on an expanded notification.
    static let maximumActionCount = 4

    /// Reached through `current()` on each use rather than stored: the class is not
    /// `Sendable`, and the seam that matters for testing is `LabelPromptDelivering` itself.
    private var center: UNUserNotificationCenter {
        .current()
    }

    func isAuthorized() async -> Bool {
        switch await center.notificationSettings().authorizationStatus {
        case .authorized, .provisional, .ephemeral: true
        case .denied, .notDetermined: false
        @unknown default: false
        }
    }

    func deliver(_ prompt: TraceLabelPrompt) async {
        // Registered here rather than at launch so the whole feature is one file to delete.
        // `setNotificationCategories` replaces the set, which is safe only while this is
        // the app's sole category — the M3 candidate notification will need a single
        // registration point, and this is the line that has to move then.
        center.setNotificationCategories([Self.category])

        let content = UNMutableNotificationContent()
        content.title = TraceLabelPrompt.title
        content.body = prompt.body()
        content.categoryIdentifier = TraceLabelPromptAction.categoryIdentifier
        // The session id is the only payload. A UUID, never a place.
        content.userInfo = [TraceLabelPromptAction.sessionIdKey: prompt.sessionId.uuidString]
        content.interruptionLevel = .active

        let request = UNNotificationRequest(
            // One notification per session: a second close cannot overwrite the first, and
            // the tap handler knows which trace it is answering about.
            identifier: "\(TraceLabelPromptAction.categoryIdentifier).\(prompt.sessionId.uuidString)",
            content: content,
            trigger: nil
        )
        do {
            try await center.add(request)
        } catch {
            // Best-effort: the trace is already on disk and the in-app screen still works.
            let nsError = error as NSError
            AppLog.detection.error(
                "trace label prompt failed: \(nsError.domain, privacy: .public)(\(nsError.code, privacy: .public))"
            )
        }
    }

    private static var category: UNNotificationCategory {
        let actions = TraceLabelPrompt.offeredModes.prefix(maximumActionCount).map { mode in
            UNNotificationAction(
                identifier: TraceLabelPromptAction.identifier(for: mode),
                title: mode.promptActionTitle,
                options: []
            )
        }
        return UNNotificationCategory(
            identifier: TraceLabelPromptAction.categoryIdentifier,
            actions: Array(actions),
            intentIdentifiers: [],
            options: []
        )
    }
}

/// Turns a tapped action into a stored label.
///
/// The notification arrives in a process that may have been launched for it, so this is
/// wired in `DetectionRuntime.bootstrap` and holds the store directly: there is no view
/// model and no screen involved in a one-tap label.
final class TraceLabelPromptResponder: NSObject, UNUserNotificationCenterDelegate {
    private let store: any TraceStoring

    init(store: any TraceStoring) {
        self.store = store
    }

    /// The pure half, so the mapping from an action identifier to a written label is
    /// testable without a `UNNotificationResponse` (which cannot be constructed).
    ///
    /// @return whether a label was written. `false` for the body tap — that opens the app,
    /// where the labelling screen takes over — and for a session the store no longer has.
    @discardableResult
    func applyLabel(actionIdentifier: String, sessionId: UUID?) -> Bool {
        guard let sessionId,
              let mode = TraceLabelPromptAction.mode(forActionIdentifier: actionIdentifier)
        else { return false }
        do {
            try store.updateLabel(TraceLabelPrompt.label(for: mode), for: sessionId)
            return true
        } catch {
            // A session evicted by the rolling cap between the prompt and the tap is an
            // ordinary outcome, not a failure worth breaking the callback for.
            AppLog.detection.notice("trace label prompt tap could not be applied: \(String(describing: error))")
            return false
        }
    }

    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {
        let userInfo = response.notification.request.content.userInfo
        let sessionId = (userInfo[TraceLabelPromptAction.sessionIdKey] as? String).flatMap(UUID.init(uuidString:))
        applyLabel(actionIdentifier: response.actionIdentifier, sessionId: sessionId)
        completionHandler()
    }
}

File: android/app/src/main/java/kr/parkingpin/app/trace/TraceLabelPromptNotifier.kt
package kr.parkingpin.app.trace

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import kr.parkingpin.app.R
import kr.parkingpin.app.data.DetectionStateStore
import kr.parkingpin.app.domain.trace.TraceLabelPrompt
import kr.parkingpin.app.domain.trace.TraceMode

/**
 * What [TraceRecorder] is allowed to know about prompting for a label.
 *
 * One method, no return value, never throws: the closed session is already on disk when
 * this is called, and a notification that cannot be posted must not be able to break the
 * recording path (docs/05_CROSS_PLATFORM_DOMAIN_CONTRACT.md §9 best-effort).
 */
interface TraceLabelPrompting {
    /** Best-effort. Deciding that permission is missing is the implementation's business. */
    suspend fun requestPrompt(prompt: TraceLabelPrompt)
}

/** Used wherever prompting is not wired, so a collaborator is never null. */
object NoOpTraceLabelPrompting : TraceLabelPrompting {
    override suspend fun requestPrompt(prompt: TraceLabelPrompt) = Unit
}

/**
 * The notification system, reduced to the two things prompting needs.
 *
 * Exists so the decision in [TraceLabelPrompter] — post, or count a suppression — is
 * testable. `NotificationManager` cannot be reached from a JVM unit test.
 */
interface LabelPromptDelivering {
    /**
     * Whether a notification would be shown at all. `false` covers a denied
     * `POST_NOTIFICATIONS`, notifications switched off for the app, and a blocked channel;
     * none of them is an error and none is retried.
     */
    fun isAuthorized(): Boolean

    /** Posts. Swallows its own failures — see [TraceLabelPrompting]. */
    fun post(prompt: TraceLabelPrompt)
}

/**
 * Decides whether a closed session is worth a notification, and counts it when it is not.
 *
 * The decision is the whole of this class, which is why it is separate from the delivery it
 * drives: a missing permission has to be *visible* rather than silent, because a field
 * weekend that collected nothing would otherwise look exactly like a weekend nobody
 * travelled (§9 "조용히 버리지 마라"). The counter is persisted for the same reason the drop
 * counters are — the prompt is posted from a process that dies between two PendingIntent
 * deliveries, so an in-memory count would always read zero by the time anyone looked.
 */
class TraceLabelPrompter(
    private val delivery: LabelPromptDelivering,
    private val stateStore: DetectionStateStore,
) : TraceLabelPrompting {

    override suspend fun requestPrompt(prompt: TraceLabelPrompt) {
        if (!delivery.isAuthorized()) {
            stateStore.addTraceLabelPromptSuppressed(1)
            return
        }
        delivery.post(prompt)
    }
}

/**
 * The names the notification and the tap that comes back have to agree on.
 *
 * **Its own channel**, never docs/04_ANDROID_IMPLEMENTATION.md §9's `parking_detection` or
 * `parking_status`. This is diagnostics: it asks about a trip that already ended, and a
 * prompt sharing a channel with a product notification would make "the user muted the wrong
 * one" unanswerable — and would silence real parking alerts the day someone gets tired of
 * being asked what the bus was.
 */
object TraceLabelPromptChannel {
    const val ID: String = "trace_label_prompt_diagnostics"
    const val ACTION_PREFIX: String = "kr.parkingpin.app.TRACE_LABEL."
    const val EXTRA_SESSION_ID: String = "kr.parkingpin.app.extra.TRACE_SESSION_ID"
    const val EXTRA_MODE: String = "kr.parkingpin.app.extra.TRACE_MODE"

    /**
     * Android shows at most three action buttons on a notification, so the tail of
     * [TraceLabelPrompt.OFFERED_MODES] is dropped rather than silently ignored by the OS.
     * iOS fits four; the label written to disk is identical either way, and the mode that
     * drops off is the one still reachable from the in-app screen.
     */
    const val MAX_ACTIONS: Int = 3

    fun actionFor(mode: TraceMode): String = ACTION_PREFIX + mode.name

    /** `null` for an intent that is not one of ours. */
    fun modeOf(action: String?): TraceMode? {
        if (action == null || !action.startsWith(ACTION_PREFIX)) return null
        val name = action.removePrefix(ACTION_PREFIX)
        return TraceMode.entries.firstOrNull { it.name == name }
    }

    /** A stable id per session, so the right notification is the one that goes away. */
    fun notificationId(sessionId: String): Int = sessionId.hashCode()
}

/**
 * Posts through [NotificationManagerCompat].
 *
 * The channel is created here rather than in `Application.onCreate` so the whole feature is
 * one file to delete; `createNotificationChannel` is idempotent and costs nothing on a
 * channel that already exists.
 *
 * `IMPORTANCE_DEFAULT` on purpose: the point is that the user notices while they still
 * remember the trip. `LOW` would file it silently into the shade, which is where the in-app
 * screen already was.
 */
class NotificationLabelPromptDelivery(context: Context) : LabelPromptDelivering {

    private val appContext: Context = context.applicationContext

    override fun isAuthorized(): Boolean =
        NotificationManagerCompat.from(appContext).areNotificationsEnabled()

    override fun post(prompt: TraceLabelPrompt) {
        try {
            createChannel()
            val builder = NotificationCompat.Builder(appContext, TraceLabelPromptChannel.ID)
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentTitle(TraceLabelPrompt.TITLE)
                .setContentText(prompt.body())
                .setStyle(NotificationCompat.BigTextStyle().bigText(prompt.body()))
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                // The body opens the app, where the labelling screen holds the modes that
                // did not fit as actions.
                .setContentIntent(openAppIntent())
                .setAutoCancel(true)

            for (mode in TraceLabelPrompt.OFFERED_MODES.take(TraceLabelPromptChannel.MAX_ACTIONS)) {
                builder.addAction(
                    0,
                    TraceLabelPrompt.actionTitle(mode),
                    labelIntent(prompt.sessionId, mode),
                )
            }

            NotificationManagerCompat.from(appContext)
                .notify(TraceLabelPromptChannel.notificationId(prompt.sessionId), builder.build())
        } catch (denied: SecurityException) {
            // POST_NOTIFICATIONS revoked between [isAuthorized] and here. Caught explicitly
            // rather than through the clause below, because it is the one failure that is
            // ordinary rather than unexpected — and because a notify() with no visible
            // SecurityException handler is a permission bug everywhere else in this app.
            Log.i(TAG, "trace label prompt not permitted: ${denied.javaClass.simpleName}")
        } catch (error: Exception) {
            // Best-effort: the trace is already on disk and the in-app screen still works.
            // The class name only, for the same reason the recording path logs no message.
            Log.e(TAG, "trace label prompt failed: ${error.javaClass.simpleName}")
        }
    }

    private fun createChannel() {
        val channel = NotificationChannel(
            TraceLabelPromptChannel.ID,
            appContext.getString(R.string.trace_label_prompt_channel_name),
            NotificationManager.IMPORTANCE_DEFAULT,
        ).apply {
            description = appContext.getString(R.string.trace_label_prompt_channel_description)
            setShowBadge(false)
        }
        appContext.getSystemService(NotificationManager::class.java)?.createNotificationChannel(channel)
    }

    /**
     * `data` as well as a distinct action, because [Intent.filterEquals] ignores extras: two
     * PendingIntents differing only in their session id would otherwise be the same one, and
     * the second trip's tap would label the first trip's session.
     */
    private fun labelIntent(sessionId: String, mode: TraceMode): PendingIntent {
        val intent = Intent(appContext, TraceLabelPromptReceiver::class.java).apply {
            action = TraceLabelPromptChannel.actionFor(mode)
            data = Uri.fromParts("parkingkok-trace-label", "$sessionId/${mode.name}", null)
            putExtra(TraceLabelPromptChannel.EXTRA_SESSION_ID, sessionId)
            putExtra(TraceLabelPromptChannel.EXTRA_MODE, mode.name)
        }
        return PendingIntent.getBroadcast(
            appContext,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun openAppIntent(): PendingIntent {
        val intent = appContext.packageManager.getLaunchIntentForPackage(appContext.packageName)
        return PendingIntent.getActivity(
            appContext,
            0,
            intent ?: Intent(),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private companion object {
        const val TAG = "ParkingpinTrace"
    }
}
